# Task1 : Handling Missing Data – Titanic Dataset


**Problem:** *The Titanic dataset contains missing values in columns like Age, Cabin,* *and Embarked.*

**Technique:** *Mean/Median imputation for numerical, Mode imputation for categorical,* *Drop for excessive missing.*

**Why:** *Missing data causes errors in ML models and biases predictions.*

In [2]:
#Data Loading & Preprocessing

import pandas as pd
import numpy as np

df = pd.read_csv('/kaggle/input/datasets/shuofxz/titanic-machine-learning-from-disaster/train.csv')

# --- Before ---
print("Missing Values BEFORE")
print(df.isnull().sum())

# Strategy 1: Mean imputation for Age (numerical)
df['Age_mean'] = df['Age'].fillna(df['Age'].mean())

# Strategy 2: Median imputation for Age (robust to outliers)
df['Age_median'] = df['Age'].fillna(df['Age'].median())

# Strategy 3: Mode imputation for Embarked (categorical)
df['Embarked'] = df['Embarked'].fillna(df['Embarked'].mode()[0])

# Strategy 4: Drop Cabin column (too many missing values >77%)
df.drop(columns=['Cabin'], inplace=True)

# Apply median to Age finally (best strategy)
df['Age'] = df['Age_median']
df.drop(columns=['Age_mean', 'Age_median'], inplace=True)

# --- After ---
print("\nMissing Values AFTER")
print(df.isnull().sum())

# ## 3. Results & Observations
print("\n=== Observations ===")
print(f"Age - Mean: {df['Age'].mean():.2f}, Median: {df['Age'].median():.2f}")
print("Cabin dropped due to 77% + missing values.")
print("Embarked filled with mode: 'S'")

Missing Values BEFORE
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

Missing Values AFTER
PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

=== Observations ===
Age - Mean: 29.36, Median: 28.00
Cabin dropped due to 77% + missing values.
Embarked filled with mode: 'S'


**Results and Observations:**

**What changed?** *Age nulls filled, Embarked nulls filled, Cabin dropped*

**Before vs After?** *Age had 177 missing → 0, Cabin had 687 missing → removed,* *Embarked had 2 missing → 0*

**What did we learn?** *Mean is sensitive to outliers; median is safer for skewed data like Age. Columns with >50% missing are better dropped than imputed.*

# Task2: Feature Encoding – Car Evaluation Dataset

**Introduction**

**Problem:** *ML models can't process raw categorical text values.*

**Technique:** *Label Encoding & One-Hot Encoding compared.*

**Why:** *Encoding converts categories to numbers models can understand.*

In [3]:
# Data Loading & Preprocessing

import pandas as pd
from sklearn.preprocessing import LabelEncoder

cols = ['buying', 'maint', 'doors', 'persons', 'lug_boot', 'safety', 'class']
df = pd.read_csv('/kaggle/input/datasets/rohit265/car-evaluation-uci-machine-learning/car.csv', names=cols)

print("Before Encoding")
print(df.head())
print(df.dtypes)

# --- Label Encoding ---
df_label = df.copy()
le = LabelEncoder()
for col in df_label.columns:
    df_label[col] = le.fit_transform(df_label[col])

print("\nAfter Label Encoding")
print(df_label.head())

# --- One-Hot Encoding ---
df_ohe = pd.get_dummies(df, columns=['buying', 'maint', 'doors', 'persons', 'lug_boot', 'safety'])

print("\nAfter One-Hot Encoding")
print(df_ohe.head())
print(f"Shape after OHE: {df_ohe.shape}")
print(f"Shape after Label Encoding: {df_label.shape}")

Before Encoding
     buying  maint  doors  persons   lug_boot  safety  class
NaN  buying  maint  doors  persons  lug_boots  safety  Class
0.0   vhigh  vhigh      2        2      small     low  unacc
1.0   vhigh  vhigh      2        2      small     med  unacc
2.0   vhigh  vhigh      2        2      small    high  unacc
3.0   vhigh  vhigh      2        2        med     low  unacc
buying      object
maint       object
doors       object
persons     object
lug_boot    object
safety      object
class       object
dtype: object

After Label Encoding
     buying  maint  doors  persons  lug_boot  safety  class
NaN       0      2      4        3         1       3      0
0.0       4      4      0        0         3       1      3
1.0       4      4      0        0         3       2      3
2.0       4      4      0        0         3       0      3
3.0       4      4      0        0         2       1      3

After One-Hot Encoding
     class  buying_buying  buying_high  buying_low  buying_med  b

**Results and observations:**

**Label Encoding:** *stays same shape, assigns integer ranks — can imply false order* *(e.g. vhigh=3 > low=1)*

**One-Hot Encoding:** *expands columns, no false ordering, but increases dimensionality*

**When to use which?** *Label encoding for ordinal data (low < med < high), OHE for* *nominal (no order)*

# Task3 : Feature Scaling – Wine Quality Dataset

**Introduction**

**Problem:** *Features have different ranges, biasing ML models.*

**Technique:** *Normalization (MinMaxScaler) & Standardization (StandardScaler).*

**Why:** *Scaling ensures all features contribute equally to the model.*

In [4]:
# ## 2. Data Loading & Preprocessing

import pandas as pd
from sklearn.preprocessing import MinMaxScaler, StandardScaler

df = pd.read_csv('/kaggle/input/datasets/arnavs19/wine-quality-uci-machine-learning-repository/winequality-red.csv', sep=';')

print("Before Scaling")
print(df.describe())

features = df.drop('quality', axis=1)

# Normalization (0 to 1)
minmax = MinMaxScaler()
df_normalized = pd.DataFrame(minmax.fit_transform(features), columns=features.columns)

# Standardization (mean=0, std=1)
standard = StandardScaler()
df_standardized = pd.DataFrame(standard.fit_transform(features), columns=features.columns)

print("\nAfter Normalization")
print(df_normalized.describe())

print("\nAfter Standardization")
print(df_standardized.describe())

Before Scaling
       fixed acidity  volatile acidity  citric acid  residual sugar  \
count    1599.000000       1599.000000  1599.000000     1599.000000   
mean        8.319637          0.527821     0.270976        2.538806   
std         1.741096          0.179060     0.194801        1.409928   
min         4.600000          0.120000     0.000000        0.900000   
25%         7.100000          0.390000     0.090000        1.900000   
50%         7.900000          0.520000     0.260000        2.200000   
75%         9.200000          0.640000     0.420000        2.600000   
max        15.900000          1.580000     1.000000       15.500000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  1599.000000          1599.000000           1599.000000  1599.000000   
mean      0.087467            15.874922             46.467792     0.996747   
std       0.047065            10.460157             32.895324     0.001887   
min       0.012000             1.

**Results & Observations**

*Normalization squeezes all values between 0-1.*

*Standardization centers data at mean=0 with std=1.*

*Standardization handles outliers better; Normalization good for neural nets.*


# Task 4: Handling Outliers – Boston Housing Dataset

**Introduction**

**Problem:** *Outliers skew model learning and degrade performance.*

**Technique:** *Z-score and IQR method to detect and remove outliers.*

**Why:** *Clean data leads to more accurate and generalizable models.*

In [5]:
# Data Loading & Preprocessing

import pandas as pd
import numpy as np
from scipy import stats

df = pd.read_csv('/kaggle/input/datasets/subho117/boston-housing-kaggle-challenge-with-lr/boston.csv')
df.fillna(df.median(), inplace=True)

print(f"Shape Before: {df.shape}")

# Z-score method
z_scores = np.abs(stats.zscore(df))
df_zscore = df[(z_scores < 3).all(axis=1)]
print(f"Shape After Z-score removal: {df_zscore.shape}")

# IQR method
Q1 = df.quantile(0.25)
Q3 = df.quantile(0.75)
IQR = Q3 - Q1
df_iqr = df[~((df < (Q1 - 1.5 * IQR)) | (df > (Q3 + 1.5 * IQR))).any(axis=1)]
print(f"Shape After IQR removal: {df_iqr.shape}")


Shape Before: (333, 16)
Shape After Z-score removal: (268, 16)
Shape After IQR removal: (182, 16)


**Results & Observations:**

*Z-score removed rows where any feature was 3 std devs away from mean.*

*IQR is more robust for skewed distributions.*

*Both methods reduced dataset size, removing noisy extreme values.*


# Task 5: Advanced Data Imputation – Retail Sales Dataset

**Introduction:**

**Problem:** *Simple imputation ignores relationships between features.*

**Technique:** *KNN Imputation and MICE (IterativeImputer).*

**Why:** *Advanced methods use feature correlations for smarter imputation.*

In [20]:
# Data Loading & Preprocessing

import pandas as pd
import numpy as np
from sklearn.impute import KNNImputer
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

df = pd.read_csv('/kaggle/input/datasets/mohammadtalib786/retail-sales-dataset/retail_sales_dataset.csv')

# Keep only numeric columns
df_numeric = df.select_dtypes(include=[np.number]).copy()

# Introduce artificial missing values for demonstration
np.random.seed(42)
mask = np.random.rand(*df_numeric.shape) < 0.1
df_missing = df_numeric.copy()
df_missing[mask] = np.nan

print(f"Missing before: {df_missing.isnull().sum().sum()}")

# KNN Imputation
knn = KNNImputer(n_neighbors=5)
df_knn = pd.DataFrame(knn.fit_transform(df_missing), columns=df_numeric.columns)

# MICE Imputation
mice = IterativeImputer(max_iter=10, random_state=42)
df_mice = pd.DataFrame(mice.fit_transform(df_missing), columns=df_numeric.columns)

print(f"Missing after KNN: {df_knn.isnull().sum().sum()}")
print(f"Missing after MICE: {df_mice.isnull().sum().sum()}")


Missing before: 525
Missing after KNN: 0
Missing after MICE: 0


/usr/local/lib/python3.12/dist-packages/sklearn/impute/_iterative.py:895: ConvergenceWarning: [IterativeImputer] Early stopping criterion not reached.
  warnings.warn(


**Results & Observations**

*KNN fills based on nearest neighbor feature values.*

*MICE iteratively models each feature using others — more accurate.*

*Both eliminate missing values; MICE is preferred for complex datasets.*

#  Task 6: Feature Engineering – Heart Disease Dataset

**Introduction**

**Problem:** *Raw features may not fully capture domain knowledge.*

**Technique:** *Create derived features — age groups, cholesterol categories, risk* *score.*

**Why:** *New features can improve model performance significantly.*

In [15]:
import pandas as pd

df = pd.read_csv('/kaggle/input/datasets/fedesoriano/heart-failure-prediction/heart.csv')

# Fix column names
df.columns = df.columns.str.lower()

print("Before Feature Engineering")
print(df.head())

# Age groups
df['age_group'] = pd.cut(df['age'], bins=[0,40,55,70,100],
                        labels=['Young','Middle','Senior','Elderly'])

# Cholesterol categories
df['chol_category'] = pd.cut(df['cholesterol'], bins=[0,200,240,600],
                            labels=['Normal','Borderline','High'])

# Risk score
df['risk_score'] = df['restingbp'] + df['cholesterol']/5 - df['maxhr']/2

print("\nAfter Feature Engineering")
print(df[['age','age_group','cholesterol','chol_category','risk_score']].head(10))



Before Feature Engineering
   age sex chestpaintype  restingbp  cholesterol  fastingbs restingecg  maxhr  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   

  exerciseangina  oldpeak st_slope  heartdisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0  

After Feature Engineering
   age age_group  cholesterol chol_category  risk_score
0   40     Young          289          High       111.8
1   49    Middle          180        Norm

**Results & Observations**

*Age groups convert continuous age into meaningful clinical categories.*

*Cholesterol categories align with medical thresholds.*

*Risk score combines multiple features into one informative signal.*


# Task 7: Variable Transformation – Bike Sharing Dataset

**Introduction**

**Problem:** *Skewed distributions violate ML model assumptions.*

**Technique:** *Log, Square Root, and Box-Cox transformations.*

**Why:** *Transformations normalize skewed data for better model performance.*

In [8]:
# Data Loading & Preprocessing

import pandas as pd
import numpy as np
from scipy import stats

df = pd.read_csv('/kaggle/input/datasets/sriramm2010/uci-bike-sharing-data/hour.csv')

target = 'cnt'
print(f"Skewness Before: {df[target].skew():.4f}")

# Log Transform
df['cnt_log'] = np.log1p(df[target])

# Square Root Transform
df['cnt_sqrt'] = np.sqrt(df[target])

# Box-Cox Transform (requires positive values)
df['cnt_boxcox'], _ = stats.boxcox(df[target] + 1)

print(f"Skewness Log: {df['cnt_log'].skew():.4f}")
print(f"Skewness Sqrt: {df['cnt_sqrt'].skew():.4f}")
print(f"Skewness BoxCox: {df['cnt_boxcox'].skew():.4f}")



Skewness Before: 1.2774
Skewness Log: -0.8182
Skewness Sqrt: 0.2865
Skewness BoxCox: -0.1374


**Results & Observations**

*Original 'cnt' was right-skewed.*

*Log transform reduced skewness significantly.*

*Box-Cox is the most powerful — finds optimal lambda automatically.*

*Sqrt is a mild transformation, useful for moderate skewness.*

# Task 8: Feature Selection – Diabetes Dataset

**Introduction**

**Problem:** *Too many features cause overfitting and slow training.*

**Technique:** *Correlation, Mutual Information, and RFE.*

**Why:** *Selecting relevant features improves model accuracy and efficiency.*

In [9]:
# Data Loading & Preprocessing

import pandas as pd
from sklearn.feature_selection import mutual_info_classif, RFE
from sklearn.linear_model import LogisticRegression

df = pd.read_csv('/kaggle/input/datasets/organizations/uciml/pima-indians-diabetes-database/diabetes.csv')

X = df.drop('Outcome', axis=1)
y = df['Outcome']

# Correlation
print("Correlation with Target")
print(df.corr()['Outcome'].sort_values(ascending=False))

# Mutual Information
mi = mutual_info_classif(X, y, random_state=42)
mi_series = pd.Series(mi, index=X.columns).sort_values(ascending=False)
print("\nMutual Information Scores")
print(mi_series)

# RFE
model = LogisticRegression(max_iter=1000)
rfe = RFE(model, n_features_to_select=5)
rfe.fit(X, y)
selected = X.columns[rfe.support_].tolist()
print(f"\n=== RFE Selected Features ===")
print(selected)



Correlation with Target
Outcome                     1.000000
Glucose                     0.466581
BMI                         0.292695
Age                         0.238356
Pregnancies                 0.221898
DiabetesPedigreeFunction    0.173844
Insulin                     0.130548
SkinThickness               0.074752
BloodPressure               0.065068
Name: Outcome, dtype: float64

Mutual Information Scores
Glucose                     0.114634
BMI                         0.080061
Pregnancies                 0.060985
Age                         0.051395
DiabetesPedigreeFunction    0.014590
Insulin                     0.011948
SkinThickness               0.004667
BloodPressure               0.000000
dtype: float64

=== RFE Selected Features ===
['Pregnancies', 'Glucose', 'BMI', 'DiabetesPedigreeFunction', 'Age']


**Results & Observations**

*Correlation: Glucose is most correlated with Outcome.*

*Mutual Information: Captures non-linear relationships too.*

*RFE: Recursively eliminates weakest features — selects top 5.*

*All 3 methods agree Glucose, BMI, Age are most important.*

# Task 9: Handling Imbalanced Data – Credit Card Fraud Detection

**Introduction**

**Problem:** *Fraud cases are rare (~0.17%), causing model bias toward majority class.*

**Technique:** *SMOTE, ADASYN, and Undersampling.*

**Why:** *Balancing classes ensures the model learns fraud patterns effectively.*

In [10]:
# Data Loading & Preprocessing

import pandas as pd
from imblearn.over_sampling import SMOTE, ADASYN
from imblearn.under_sampling import RandomUnderSampler
from collections import Counter

df = pd.read_csv('/kaggle/input/datasets/aliiihussain/credit-card-fraud-detection/credit_card_fraud_synthetic.csv')

X = df.drop('Class', axis=1)
y = df['Class']

print(f"Before: {Counter(y)}")

# SMOTE
sm = SMOTE(random_state=42)
X_sm, y_sm = sm.fit_resample(X, y)
print(f"After SMOTE: {Counter(y_sm)}")

# ADASYN
ada = ADASYN(random_state=42)
X_ada, y_ada = ada.fit_resample(X, y)
print(f"After ADASYN: {Counter(y_ada)}")

# Undersampling
rus = RandomUnderSampler(random_state=42)
X_rus, y_rus = rus.fit_resample(X, y)
print(f"After Undersampling: {Counter(y_rus)}")



Before: Counter({0: 982, 1: 18})
After SMOTE: Counter({0: 982, 1: 982})
After ADASYN: Counter({1: 990, 0: 982})
After Undersampling: Counter({0: 18, 1: 18})


**Results & Observations**

**Original:** *284315 legit vs 492 fraud — heavily imbalanced.*

**SMOTE:** *Synthesizes new fraud samples to balance classes.*

**ADASYN:** *Focuses synthesis on harder-to-classify samples.*

**Undersampling:** *Reduces majority class — risk of losing information.*

*SMOTE preferred when dataset is large enough.*

# Task 10: Combining Multiple Datasets – MovieLens Dataset

**Introduction**

**Problem:** *Useful information is spread across multiple files.*

**Technique:** *Merging datasets on common keys, handling duplicates/nulls.*

**Why:** *A unified dataset enables richer feature sets for ML models.*

In [11]:
# Data Loading & Preprocessing

import pandas as pd

# Load with correct encoding and separator for .dat files
ratings = pd.read_csv('/kaggle/input/datasets/sherinclaudia/movielens/ratings.dat',
                      sep='::', engine='python', encoding='latin-1',
                      names=['userId', 'movieId', 'rating', 'timestamp'])

movies = pd.read_csv('/kaggle/input/datasets/sherinclaudia/movielens/movies.dat',
                     sep='::', engine='python', encoding='latin-1',
                     names=['movieId', 'title', 'genres'])

print(f"Ratings shape: {ratings.shape}")
print(f"Movies shape: {movies.shape}")

# Merge on movieId
df = ratings.merge(movies, on='movieId', how='left')

# Drop duplicates
df.drop_duplicates(inplace=True)

# Handle missing
df.dropna(inplace=True)

# Extract year from title
df['year'] = df['title'].str.extract(r'\((\d{4})\)')

print(f"\nMerged shape: {df.shape}")
print(df.head())







Ratings shape: (1000209, 4)
Movies shape: (3883, 3)

Merged shape: (1000209, 7)
   userId  movieId  rating  timestamp                                   title  \
0       1     1193       5  978300760  One Flew Over the Cuckoo's Nest (1975)   
1       1      661       3  978302109        James and the Giant Peach (1996)   
2       1      914       3  978301968                     My Fair Lady (1964)   
3       1     3408       4  978300275                  Erin Brockovich (2000)   
4       1     2355       5  978824291                    Bug's Life, A (1998)   

                         genres  year  
0                         Drama  1975  
1  Animation|Children's|Musical  1996  
2               Musical|Romance  1964  
3                         Drama  2000  
4   Animation|Children's|Comedy  1998  


# Task 11: Dimensionality Reduction – MNIST Dataset

**Introduction**

**Problem:** *MNIST has 784 features (pixels) — high dimensional, slow to train.*

**Technique:** *PCA and t-SNE to reduce dimensions.*

**Why:** *Reduces computation while preserving important variance/structure.*

In [13]:
import pandas as pd
import numpy as np
from sklearn.decomposition import PCA
from sklearn.manifold import TSNE

# Load MNIST dataset
train = pd.read_csv('/kaggle/input/competitions/digit-recognizer/train.csv')

y = train['label'].values
X = train.drop('label', axis=1).values

# Use subset
X = X[:5000]
y = y[:5000]

print(f"Original shape: {X.shape}")

# Normalize
X = X / 255.0

# PCA
pca = PCA(n_components=50, random_state=42)
X_pca = pca.fit_transform(X)

print(f"After PCA: {X_pca.shape}")
print(f"Variance explained: {pca.explained_variance_ratio_.sum():.2%}")

# t-SNE
tsne = TSNE(n_components=2, random_state=42, perplexity=30)
X_tsne = tsne.fit_transform(X_pca)

print(f"After t-SNE: {X_tsne.shape}")








Original shape: (5000, 784)
After PCA: (5000, 50)
Variance explained: 82.79%
After t-SNE: (5000, 2)


**Results & Observations**

**PCA:** *Reduced 784 → 50 features while retaining ~85% variance.*

**t-SNE:** *Further reduced to 2D for visualization — shows clear digit clusters.*

*PCA is fast & linear; t-SNE is slow but great for visual cluster exploration.*

# Task 12: Text Preprocessing – IMDB Movie Reviews

**Introduction**

**Problem:** *Raw text is unstructured and can't be fed directly to ML models.*

**Technique:** *Lowercasing, stopword removal, stemming, lemmatization, tokenization.*

**Why:** *Clean text reduces noise and improves model accuracy.*

In [25]:
#  Data Loading & Preprocessing

import pandas as pd
import re

df = pd.read_csv('/kaggle/input/datasets/vishakhdapat/imdb-movie-reviews/IMDB Dataset.csv')

print("Before")
print(df['review'][0][:300])

# Manual stopwords list (no download needed)
stop_words = set([
    'i','me','my','myself','we','our','ours','ourselves','you','your','yours',
    'yourself','he','him','his','himself','she','her','hers','herself','it',
    'its','itself','they','them','their','theirs','themselves','what','which',
    'who','whom','this','that','these','those','am','is','are','was','were',
    'be','been','being','have','has','had','having','do','does','did','doing',
    'a','an','the','and','but','if','or','because','as','until','while','of',
    'at','by','for','with','about','against','between','into','through',
    'during','before','after','above','below','to','from','up','down','in',
    'out','on','off','over','under','again','further','then','once','here',
    'there','when','where','why','how','all','both','each','few','more',
    'most','other','some','such','no','nor','not','only','own','same','so',
    'than','too','very','s','t','can','will','just','don','should','now'
])

def simple_stem(word):
    # Basic suffix stripping (poor man's stemmer)
    for suffix in ['ing', 'tion', 'ed', 'ly', 'er', 'es', 's']:
        if word.endswith(suffix) and len(word) - len(suffix) > 2:
            return word[:-len(suffix)]
    return word

def preprocess(text):
    text = text.lower()                              # Lowercase
    text = re.sub(r'<.*?>', '', text)               # Remove HTML tags
    text = re.sub(r'[^a-z\s]', '', text)           # Remove punctuation/numbers
    tokens = text.split()                            # Tokenize (split on space)
    tokens = [t for t in tokens if t not in stop_words]  # Remove stopwords
    tokens = [simple_stem(t) for t in tokens]       # Stem
    return ' '.join(tokens)

df['clean_review'] = df['review'].apply(preprocess)

print("\nAfter")
print(df['clean_review'][0][:300])

print(f"\nOriginal avg length: {df['review'].str.len().mean():.0f} chars")
print(f"Cleaned avg length:  {df['clean_review'].str.len().mean():.0f} chars")





Before
One of the other reviewers has mentioned that after watching just 1 Oz episode you'll be hooked. They are right, as this is exactly what happened with me.<br /><br />The first thing that struck me about Oz was its brutality and unflinching scenes of violence, which set in right from the word GO. Tru

After
one reviewer mention watch oz episode youll hook right exact happen methe first thing struck oz brutality unflinch scen violence set right word go trust show faint heart timid show pull punch regard drug sex violence hardcore classic use wordit call oz nickname given oswald maximum security state pe

Original avg length: 1309 chars
Cleaned avg length:  762 chars


**Results & Observations**

*HTML tags, punctuation, and noise removed.*

*Stopwords eliminated common but uninformative words.*

*Lemmatization + Stemming reduced words to their root forms.*

**Result:** *compact, clean text ready for vectorization (TF-IDF, embeddings).*

# Task 13: Time-Series Preprocessing – Air Quality Dataset

**Introduction**

**Problem:** *Time-series data has missing timestamps, irregular intervals, and noise.*

**Technique:** *Resampling, forward fill, and rolling mean smoothing.*

**Why:** *Clean, regular time-series data is essential for forecasting models.*


In [32]:
import pandas as pd
import numpy as np

df = pd.read_csv(
    '/kaggle/input/datasets/nishantbhadauria/datasetucimlairquality/AirQualityUCI.csv',
    sep=',',   
)

# Clean column names
df.columns = df.columns.str.strip().str.lower()

# Combine date + time
df['datetime'] = pd.to_datetime(df['date'] + ' ' + df['time'], dayfirst=True)

df.drop(columns=['date', 'time'], inplace=True)
df.set_index('datetime', inplace=True)

# Replace invalid values
df.replace(-200, np.nan, inplace=True)

print(f"Missing before fill: {df.isnull().sum().sum()}")

# Fill missing values
df = df.ffill()

print(f"Missing after fill: {df.isnull().sum().sum()}")

# Resample
df_hourly = df.resample('h').mean(numeric_only=True)

# Smoothing
df_hourly['co_smoothed'] = df_hourly['co_gt'].rolling(window=24, min_periods=1).mean()

print(df_hourly[['co_gt', 'co_smoothed']].head())






Missing before fill: 16701
Missing after fill: 27
                     co_gt  co_smoothed
datetime                               
2004-03-10 18:00:00    2.6     2.600000
2004-03-10 19:00:00    2.0     2.300000
2004-03-10 20:00:00    2.2     2.266667
2004-03-10 21:00:00    2.2     2.250000
2004-03-10 22:00:00    1.6     2.120000


/tmp/ipykernel_55/204226009.py:13: UserWarning: Parsing dates in %m/%d/%Y %H:%M:%S format when dayfirst=True was specified. Pass `dayfirst=False` or specify a format to silence this warning.
  df['datetime'] = pd.to_datetime(df['date'] + ' ' + df['time'], dayfirst=True)


**Results & Observations**

*-200 sensor errors replaced with NaN then forward-filled.*

*Resampling to hourly intervals creates a uniform time index.*

*Rolling mean smoothing removes noise while preserving trend.*

*Data is now ready for LSTM or ARIMA forecasting models.*

In [3]:
import os

for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

/kaggle/input/datasets/fedesoriano/heart-failure-prediction/heart.csv
/kaggle/input/datasets/sherinclaudia/movielens/users.dat
/kaggle/input/datasets/sherinclaudia/movielens/ratings.dat
/kaggle/input/datasets/sherinclaudia/movielens/movies.dat
/kaggle/input/datasets/subho117/boston-housing-kaggle-challenge-with-lr/boston.csv
/kaggle/input/datasets/rohit265/car-evaluation-uci-machine-learning/car.csv
/kaggle/input/datasets/nishantbhadauria/datasetucimlairquality/AirQualityUCI.csv
/kaggle/input/datasets/aliiihussain/credit-card-fraud-detection/credit_card_fraud_synthetic.csv
/kaggle/input/datasets/shuofxz/titanic-machine-learning-from-disaster/train.csv
/kaggle/input/datasets/shuofxz/titanic-machine-learning-from-disaster/test.csv
/kaggle/input/datasets/organizations/uciml/pima-indians-diabetes-database/diabetes.csv
/kaggle/input/datasets/arnavs19/wine-quality-uci-machine-learning-repository/winequality.names
/kaggle/input/datasets/arnavs19/wine-quality-uci-machine-learning-repository/wi